keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1284 @@
+
+Some third-party frameworks have implemented retry policies for their pods:
+- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
+- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))
+
+Additionally, some pod failures are not linked with the container execution,
+but rather with the internal kubernetes cluster management (see:
+[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).
+Such pod failures should be recongnized as infrastructure failures and it
+should be possible to ignore them from the counter towards `backoffLimit`.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Extension of Job API with user-friendly syntax to terminate jobs based on:
+  * container exit codes
+  * `status.reason` field of a failed pod.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Implementation of other indicators of non-retriable jobs such as termination logs.
+- Modification of the semantics for job termination. In particular, allowing for
+  all indexes of an indexed-job to execute when only one or a few indexes fail
+  [#109712](https://github.com/kubernetes/kubernetes/issues/109712).
+- Similar termination policies for other workload controllers such as Deployments
+  or StatefulSets.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Extension of the Job API with a new field which allows to configure the set of
+conditions which determine how a pod failure is handled. We extend the Job API
+to support discrimination of pod failures based on container exit codes as well
+as the pod's `status.reason` field.
+
+<!--
+The value of status.reason I found used by k8s currently is:
+	// NodeUnreachablePodReason is the reason on a pod when its state cannot be confirmed as kubelet is unresponsive
+	// on the node it is (was) running.
+NodeUnreachablePodReason = "NodeLost" (defined on node.go)
+-->
+The pod's `status.reason` field is currently defined but used in few situations.
+We extend its use to store the reason of a pod's failure when originated by
+a kubernetes component (kube-scheduler, kubelet, pod-gc, etc.). To achieve this
+we extend the Pod delete API to pass the reason for pod termination. Such
```

Comment:
I recommend not changing the delete api, since it's not reasonable to do this specially for specific apis (such as pods). Since that code is generic.

Instead just change these components to set a status condition before deleting. Since you have to change them anyway.

Reply:
1. We want to target the existing Pod.status.reason field, which kubelet already uses. And one of the kcm node controllers (an update right before calling Delete)
2. The alternative you propose is certainly an option, but it has the disadvantage of adding API calls. Preemption is already a slow process in kube-scheduler (one delete for each low priority pod that needs to be deleted). I would prefer not to add another API call. But we can certainly go this path if api-machinery strongly disagrees.

Reply:
3. I don't see why other resources could not benefit from having a reason in their status, but it's out-of-the scope of this KEP.

Reply:
Sorry, I missed that you were not talking about a condition, but that doesn't change my answer.

pod's status.reason sounds like a field that should not exist if we'd gotten conditions right to begin with. We would not want to add a way to set that field on delete on other objects.

There's a small chance you can figure out a way to do this with mutating webhooks, in which case no one can really stop you, but I still recommend against doing this. (I also think it probably won't work but I'm not sure specifically what prevents it)

I don't think it is bad to do an extra API call to both update and delete. If you are worried about QPS, don't be, I think premption is a very high-priority item and we can get you more throughput to do it if it's needed.

Reply:
BTW, I think it's a very bad idea to have a single field with multiple controllers (kubelet + controller manager + new use) that all want to set it. I think you should add a new field, or make a new condition. The api is intended to help controllers set *different parts* of the object, not to share responsibility for a single field, I'm not sure how that could work. Kubelet could clobber your write, for example.

Reply:
> I think it's a very bad idea to have a single field with multiple controllers that all want to set it.

Fair, we can probably change the proposal to use PodConditions instead. We just have to change kubelet to also do the same (which means we need sig-node as participating-sig?)

> If you are worried about QPS, don't be, I think premption is a very high-priority item and we can get you more throughput to do it if it's needed.

Sounds good. We can put "tweaking the scheduler priority" that as a beta criteria based on scalability testing.

Reply:
Discussed offline with @lavalamp:

If we are going the path of an update before a delete, then this is of no concern of sig-api-machinery and can be removed from participating sigs.

Let's add sig-node.

Reply:
:+1: for using PodCondition

Reply:
The KEP uses now `pod.status.conditions` field instead of the `status.reason` field. Also, there is a comment of applying rules when we have multiple conditions:

Note that, a pod can have multiple Pod conditions with multiple types present. A rule is applied if any of the conditions matches the specified rule requirements. Also, it is only the first rule with matching requirements which is applied as the rules are evaluated in order.